nominal/network/nominal_session.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+import gzip
+import json
+import logging
+from typing import Type, TypeVar
+
+import requests
+from conjure_python_client import RequestsClient, ServiceConfiguration
+from conjure_python_client._http.requests_client import RetryWithJitter, TransportAdapter
+
+logger = logging.getLogger(__name__)
+T = TypeVar("T")
+
+
+class NominalRequestsClient(RequestsClient):
+    """Wrapper around conjures RequestClient to automatically gzip post-requested data.
+
+    In bandwidth constrained scenarios, this has been measured to have 5-6x speedups in
+    uploading data, depending on its compressability.
+    """
+
+    @classmethod
+    def create(
+        cls,
+        service_class: Type[T],
+        user_agent: str,
+        service_config: ServiceConfiguration,
+        return_none_for_unknown_union_types: bool = False,
+        enable_keep_alive: bool = False,
+        session: requests.Session | None = None,
+    ) -> T:
+        # setup retry to match java remoting
+        # https://github.com/palantir/http-remoting/tree/3.12.0#quality-of-service-retry-failover-throttling
+        retry = RetryWithJitter(
+            total=service_config.max_num_retries,
+            read=0,  # do not retry read errors
+            status_forcelist=[308, 429, 503],
+            backoff_factor=float(service_config.backoff_slot_size) / 1000,
+        )
+        transport_adapter = TransportAdapter(max_retries=retry, enable_keep_alive=enable_keep_alive)
+        # create a session, for shared connection polling, user agent, etc
+        if session is None:
+            session = NominalRequestsSession()
+
+        session.headers["User-Agent"] = user_agent
+        if service_config.security is not None:
+            verify = service_config.security.trust_store_path
+        else:
+            verify = None
+        for uri in service_config.uris:
+            session.mount(uri, transport_adapter)
+
+        return service_class(  # type: ignore
+            session,
+            service_config.uris,
+            service_config.connect_timeout,
+            service_config.read_timeout,
+            verify,
+            return_none_for_unknown_union_types,
+        )
+
+
+class NominalRequestsSession(requests.Session):
```

Comment:
I think we should be customizing a transport adapter, not the session. Also would name it more semantically.

Reply:
tldr I want to avoid the redundancy